File: impl/java/src/test/java/com/cptnfizzbin/keycard/SubjectFieldMapperTest.java
package com.cptnfizzbin.keycard;

import lombok.AllArgsConstructor;
import lombok.Getter;
import org.junit.Test;

import com.cptnfizzbin.keycard.action.Action;
import com.cptnfizzbin.keycard.action.ActionFactory;
import com.cptnfizzbin.keycard.errors.PolicyLoadException;
import com.cptnfizzbin.keycard.policy.Policy;
import com.cptnfizzbin.keycard.policy.PolicyDefinition;
import com.cptnfizzbin.keycard.subject.Subject;
import com.cptnfizzbin.keycard.subject.SubjectFactory;
import com.cptnfizzbin.keycard.subject.SubjectFieldMapper;
import com.cptnfizzbin.keycard.subject.SubjectFieldMapperCatalog;

import java.util.List;
import java.util.Map;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertThrows;
import static org.junit.Assert.assertTrue;

public class SubjectFieldMapperTest {
    @Getter
    @AllArgsConstructor
    static class Author {
        private final String name;
    }

    @Getter
    @AllArgsConstructor
    static class Post {
        private final String status;
        private final Author author;
    }

    private static SubjectFieldMapper<Post> authorNameMapper() {
        return SubjectFieldMapper.<Post>builder()
            .field("authorName", (post) -> post.getAuthor().getName())
            .build();
    }

    @Test
    public void mappedFieldIsResolvedThroughItsGetterInsteadOfReflection() {
        Subject<Post> post = SubjectFactory.create("Post", authorNameMapper());
        Action<String> read = ActionFactory.create("Read");

        Policy policy = Policy.from(new PolicyDefinition("1.0.0", List.of(
            new PolicyDefinition.Rule("allow", "Read", "Post", Map.of("authorName", "Alice"))
        )));

        assertTrue(policy.can(read, post.wrap(new Post("draft", new Author("Alice")))));
        assertFalse(policy.can(read, post.wrap(new Post("draft", new Author("Bob")))));
    }

    @Test
    public void aFieldTheMapperDoesntDefineFallsBackToReflection() {
        Subject<Post> post = SubjectFactory.create("Post", authorNameMapper());
        Action<String> read = ActionFactory.create("Read");

        Policy policy = Policy.from(new PolicyDefinition("1.0.0", List.of(
            new PolicyDefinition.Rule("allow", "Read", "Post", Map.of("status", "draft"))
        )));

        assertTrue(policy.can(read, post.wrap(new Post("draft", new Author("Alice")))));
        assertFalse(policy.can(read, post.wrap(new Post("published", new Author("Alice")))));
    }

    @Test
    public void doesNotApplyOnceAFieldLookupHasNarrowedIntoANestedValue() {
        // The mapper is bound to the top-level Post instance - `author.name`
        // still resolves via reflection on the narrowed `author` object,
        // not through the mapper (which is never even asked about "name").
        Subject<Post> post = SubjectFactory.create("Post", authorNameMapper());
        Action<String> read = ActionFactory.create("Read");

        Policy policy = Policy.from(new PolicyDefinition("1.0.0", List.of(
            new PolicyDefinition.Rule("allow", "Read", "Post", Map.of("author", Map.of("name", "Alice")))
        )));

        assertTrue(policy.can(read, post.wrap(new Post("draft", new Author("Alice")))));
    }

    @Test
    public void stillAppliesInsideAndOrNotWhichEvaluateAgainstTheSameTopLevelSubject() {
        Subject<Post> post = SubjectFactory.create("Post", authorNameMapper());
        Action<String> read = ActionFactory.create("Read");

        Policy policy = Policy.from(new PolicyDefinition("1.0.0", List.of(
            new PolicyDefinition.Rule("allow", "Read", "Post", Map.of(
                "$and", List.of(
                    Map.of("authorName", "Alice"),
                    Map.of("$or", List.of(Map.of("status", "draft"), Map.of("authorName", "Alice")))
                )
            ))
        )));

        assertTrue(policy.can(read, post.wrap(new Post("published", new Author("Alice")))));
        assertFalse(policy.can(read, post.wrap(new Post("published", new Author("Bob")))));
    }

    @Test
    public void wrapCarriesTheFieldMapperForwardUnchanged() {
        Subject<Post> post = SubjectFactory.create("Post", authorNameMapper());
        Subject<Post> wrapped = post.wrap(new Post("draft", new Author("Alice")));

        assertEquals(post.getFieldMapper(), wrapped.getFieldMapper());
    }

    @Test
    public void catalogMapperIsConsultedWhenTheSubjectInHandCarriesNoneOfItsOwn() {
        Subject<Post> post = SubjectFactory.create("Post");
        Action<String> read = ActionFactory.create("Read");

        SubjectFieldMapperCatalog catalog = SubjectFieldMapperCatalog.builder()
            .register("Post", authorNameMapper())
            .build();
        KeycardConfig config = KeycardConfig.builder().mapper(catalog).build();

        Policy policy = Policy.from(new PolicyDefinition("1.0.0", List.of(
            new PolicyDefinition.Rule("allow", "Read", "Post", Map.of("authorName", "Alice"))
        )), config);

        assertTrue(policy.can(read, post.wrap(new Post("draft", new Author("Alice")))));
        assertFalse(policy.can(read, post.wrap(new Post("draft", new Author("Bob")))));
    }

    @Test
    public void aFieldMapperOnTheSubjectItselfTakesPrecedenceOverTheCatalog() {
        Subject<Post> post = SubjectFactory.create("Post", authorNameMapper());
        Action<String> read = ActionFactory.create("Read");

        // Deliberately mismatched, so the test can tell which one won.
        SubjectFieldMapperCatalog catalog = SubjectFieldMapperCatalog.builder()
            .register("Post", SubjectFieldMapper.<Post>builder().field("authorName", (p) -> "Mismatched").build())
            .build();
        KeycardConfig config = KeycardConfig.builder().mapper(catalog).build();

        Policy policy = Policy.from(new PolicyDefinition("1.0.0", List.of(
            new PolicyDefinition.Rule("allow", "Read", "Post", Map.of("authorName", "Alice"))
        )), config);

        assertTrue(policy.can(read, post.wrap(new Post("draft", new Author("Alice")))));
    }

    @Test
    public void configActionsAndSubjectsWidenTheEc8Catalog() {
        Subject<Post> post = SubjectFactory.create("Post");
        KeycardConfig config = KeycardConfig.builder()
            .actions(List.of(ActionFactory.create("Read")))
            .subjects(List.of(post))
            .build();

        assertThrows(PolicyLoadException.class, () ->
            Policy.from(new PolicyDefinition("1.0.0", List.of(
                new PolicyDefinition.Rule("allow", "Write", "Post", null)
            )), config));

        Policy.from(new PolicyDefinition("1.0.0", List.of(
            new PolicyDefinition.Rule("allow", "Read", "Post", null)
        )), config); // should not throw
    }
}

File: impl/js/src/subject/subjectFieldMapper.test.ts
import { describe, expect, test } from "vitest"

import { createSubject } from "./subjectFactory.ts"
import { SubjectFieldMapperCatalog } from "./subjectFieldMapperCatalog.ts"
import { createAction } from "../action/index.ts"
import { Policy } from "../policy/policy.ts"

interface Post {
  status: string
  author: { name: string }
}

const Read = createAction("Read")

describe("SubjectFieldMapper: attached via createSubject", () => {
  test("a mapped field is resolved through its getter instead of property access", () => {
    const post = createSubject<Post>("Post", {
      authorName: (instance) => instance.author.name,
    })

    const policy = Policy.from({
      version: "0.1",
      rules: [["allow", "Read", "Post", { authorName: "Alice" }]],
    })

    expect(policy.can(Read, post.wrap({ status: "draft", author: { name: "Alice" } }))).toBe(true)
    expect(policy.can(Read, post.wrap({ status: "draft", author: { name: "Bob" } }))).toBe(false)
  })

  test("a field the mapper doesn't define still falls back to ordinary property access", () => {
    const post = createSubject<Post>("Post", {
      authorName: (instance) => instance.author.name,
    })

    const policy = Policy.from({
      version: "0.1",
      rules: [["allow", "Read", "Post", { status: "draft" }]],
    })

    expect(policy.can(Read, post.wrap({ status: "draft", author: { name: "Alice" } }))).toBe(true)
    expect(policy.can(Read, post.wrap({ status: "published", author: { name: "Alice" } }))).toBe(false)
  })

  test("a field condition cannot narrow twice even with a mapper in play", () => {
    // v1 permits only one level of field narrowing -
    // a fieldMapper on Post doesn't change that: `author` is a field of
    // Post, but `author`'s own Condition can't itself be another field
    // condition (`name`), mapped or not.
    const post = createSubject<Post>("Post", {
      authorName: (instance) => instance.author.name,
    })

    const policy = Policy.from({
      version: "0.1",
      rules: [["allow", "Read", "Post", { author: { name: "Alice" } }]],
    })

    expect(policy.can(Read, post.wrap({ status: "draft", author: { name: "Alice" } }))).toBe(false)
  })

  test("a mapped field's value can still use non-field operators", () => {
    // Once resolved through the mapper, `authorName`'s own value (a plain
    // string here) can still be checked with any non-field operator -
    // narrowing is what's restricted to one level, not operator use.
    const post = createSubject<Post>("Post", {
      authorName: (instance) => instance.author.name,
    })

    const policy = Policy.from({
      version: "0.1",
      rules: [["allow", "Read", "Post", { authorName: { $substr: "Ali" } }]],
    })

    expect(policy.can(Read, post.wrap({ status: "draft", author: { name: "Alice" } }))).toBe(true)
    expect(policy.can(Read, post.wrap({ status: "draft", author: { name: "Bob" } }))).toBe(false)
  })

  test("still applies inside $and/$or/$not, which evaluate against the same top-level subject", () => {
    const post = createSubject<Post>("Post", {
      authorName: (instance) => instance.author.name,
    })

    const policy = Policy.from({
      version: "0.1",
      rules: [[
        "allow", "Read", "Post",
        { $and: [{ authorName: "Alice" }, { $or: [{ status: "draft" }, { authorName: "Alice" }] }] },
      ]],
    })

    expect(policy.can(Read, post.wrap({ status: "published", author: { name: "Alice" } }))).toBe(true)
    expect(policy.can(Read, post.wrap({ status: "published", author: { name: "Bob" } }))).toBe(false)
  })

  test("`wrap` carries the field mapper forward unchanged", () => {
    const post = createSubject<Post>("Post", { authorName: (instance) => instance.author.name })
    const wrapped = post.wrap({ status: "draft", author: { name: "Alice" } })

    expect(wrapped.fieldMapper).toBe(post.fieldMapper)
  })
})

describe("SubjectFieldMapper: registered via KeycardConfig.mapper", () => {
  test("is consulted when the Subject in hand carries no fieldMapper of its own", () => {
    const post = createSubject<Post>("Post")
    const catalog = new SubjectFieldMapperCatalog({
      Post: { authorName: (instance: Post) => instance.author.name },
    })

    const policy = Policy.from(
      {
        version: "0.1",
        rules: [["allow", "Read", "Post", { authorName: "Alice" }]],
      },
      { mapper: catalog },
    )

    expect(policy.can(Read, post.wrap({ status: "draft", author: { name: "Alice" } }))).toBe(true)
    expect(policy.can(Read, post.wrap({ status: "draft", author: { name: "Bob" } }))).toBe(false)
  })

  test("a fieldMapper on the Subject itself takes precedence over the catalog", () => {
    const post = createSubject<Post>("Post", { authorName: (instance) => instance.author.name })
    const catalog = new SubjectFieldMapperCatalog({
      // Deliberately different, so the test can tell which one won.
      Post: { authorName: () => "Mismatched" },
    })

    const policy = Policy.from(
      {
        version: "0.1",
        rules: [["allow", "Read", "Post", { authorName: "Alice" }]],
      },
      { mapper: catalog },
    )

    expect(policy.can(Read, post.wrap({ status: "draft", author: { name: "Alice" } }))).toBe(true)
  })
})

describe("KeycardConfig: actions/subjects widen meta.actions/meta.subjects (EC-8)", () => {
  test("Policy validates rule actions/subjects against config.actions/config.subjects even without a matching meta declaration", () => {
    expect(() =>
      Policy.from(
        { version: "0.1", rules: [["allow", "Write", "Post"]] },
        { actions: { Read: createAction("Read") }, subjects: { Post: createSubject("Post") } },
      ),
    ).toThrow()

    expect(() =>
      Policy.from(
        { version: "0.1", rules: [["allow", "Read", "Post"]] },
        { actions: { Read: createAction("Read") }, subjects: { Post: createSubject("Post") } },
      ),
    ).not.toThrow()
  })
})
